jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp

```diff
@@ -0,0 +1,369 @@
+  CreateAllocaMemoryObject(const rvsdg::node &allocaNode);
+
+  PointerObject::Index
+  CreateMallocMemoryObject(const rvsdg::node &mallocNode);
+
+  PointerObject::Index
+  CreateGlobalMemoryObject(const delta::node &deltaNode);
+
+  PointerObject::Index
+  CreateFunctionMemoryObject(const lambda::node &lambdaNode);
+
+  PointerObject::Index
+  CreateImportMemoryObject(const rvsdg::argument &importNode);
+
+  const std::unordered_map<const rvsdg::output *, PointerObject::Index>&
+  GetRegisterMap() const noexcept;
+
+  const std::unordered_map<const rvsdg::node *, PointerObject::Index>&
+  GetAllocaMap() const noexcept;
+
+  const std::unordered_map<const rvsdg::node *, PointerObject::Index>&
+  GetMallocMap() const noexcept;
+
+  const std::unordered_map<const delta::node *, PointerObject::Index>&
+  GetGlobalMap() const noexcept;
+
+  const std::unordered_map<const lambda::node *, PointerObject::Index>&
+  GetFunctionMap() const noexcept;
+
+  const std::unordered_map<const rvsdg::argument *, PointerObject::Index>&
+  GetImportMap() const noexcept;
+
+  PointerObject::Index
+  NumPointerObjects() const noexcept;
+
+  PointerObject &
+  GetPointerObject(PointerObject::Index index);
+
+  const PointerObject &
+  GetPointerObject(PointerObject::Index index) const;
+
+  const std::unordered_set<PointerObject::Index> &
+  GetPointsToSet(PointerObject::Index idx) const;
+
+  /**
+   * Adds pointee to P(pointer)
+   * @param pointer the index of the PointerObject that points
+   * @param pointee the index of the PointerObject that is pointed at, can not be a register.
+   * @return true if P(pointer) was changed by this operation
+   */
+  bool
+  AddToPointsToSet(PointerObject::Index pointer, PointerObject::Index pointee);
+
+  /**
+   * Makes P(superset) a superset of P(subset) by adding any elements in the set difference
+   * @param superset the index of the PointerObject that shall point to everything subset points to
+   * @param subset the index of the PointerObject whose pointees shall all be pointed to by superset as well
+   * @return true if P(superset) was modified by this operation
+   */
+  bool
+  MakePointsToSetSuperset(PointerObject::Index superset, PointerObject::Index subset);
+
+  /**
+   * Adds the Escaped flag to all PointerObjects in the P(pointer) set
+   * @param pointer the pointer whose pointees should be marked as escaped
+   * @return true if any PointerObjects had their flag modified by this operation
+   */
+  bool
+  MarkAllPointeesAsEscaped(PointerObject::Index pointer);
+};
+
+/**
+ * A constraint of the form:
+ * P(superset) is a superset of P(subset)
+ * Example of application is when a register has multiple source values
+ */
+class SupersetConstraint final
+{
+  PointerObject::Index Superset_;
+  PointerObject::Index Subset_;
+
+public:
+  SupersetConstraint(PointerObject::Index superset, PointerObject::Index subset)
+          : Superset_(superset), Subset_(subset) {}
+
+  bool
+  Apply(PointerObjectSet& set);
```

Comment:
What does this function do? Apply is a rather generic name? Some documentation might help.

Reply:
Meh, after reading it a second time, it obviously applies the constraint. Maybe still some documentation that highlights this for sleepy readers....